zerver/views/users.py

```diff
@@ -116,26 +116,30 @@ def update_user_backend(request, user_profile, email,
 # TODO: Since eventually we want to support using the same email with
 # different organizations, we'll eventually want this to be a
 # logged-in endpoint so that we can access the realm_id.
+@zulip_login_required
 def avatar(request, email_or_id, medium=False):
     # type: (HttpRequest, str, bool) -> HttpResponse
     """Accepts an email address or user ID and returns the avatar"""
+    is_email = False
     try:
         int(email_or_id)
     except ValueError:
-        get_user_func = get_user_profile_by_email  # type: Callable[..., UserProfile]
-    else:
-        get_user_func = get_user_profile_by_id
+        is_email = True
 
     try:
         # If there is a valid user account passed in, use its avatar
-        user_profile = get_user_func(email_or_id)
+        realm = request.user.realm
+        if is_email:
+            user_profile = get_user(email_or_id, realm)
+        else:
+            user_profile = get_user_profile_by_id(email_or_id)
         url = avatar_url(user_profile, medium=medium)
     except UserProfile.DoesNotExist:
         # If there is no such user, treat it as a new gravatar
         email = email_or_id
         avatar_source = 'G'
         avatar_version = 1
-        url = get_avatar_url(avatar_source, email, avatar_version, medium=medium)
+        url = get_avatar_url(avatar_source, email, avatar_version, medium=medium, realm_id=realm.id)
```

Comment:
On a related note, I think we should probably start migrating `static/js/popovers.js` to use user IDs, not emails, to generate its `/avatar/` URLs, since user IDs are much cleaner long-term.  Eventually, we'll want to deprecate the email case in this endpoint.